SQL: refactor In predicate moving it to QL project

astefan opened the pull request:
Addresses https://github.com/elastic/elasticsearch/issues/52637.

* Move tests classes to the QL project.
* Create SQL dedicated In class to handle SQL specific data types.
* Update SQL classes to use the InPipe and InProcessor QL classes.
* Clean up Foldables code and keep the bare minimum in QL project.

astefan committed:
Move In, InPipe and InProcessor out of SQL to the common QL project.
Move tests classes to the QL project.
Create SQL dedicated In class to handle SQL specific data types.
Update SQL classes to use the InPipe and InProcessor QL classes.
Extract common Foldables methods in QL project.

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 52637_impl

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Foldables.java:
```diff
@@ -0,0 +1,44 @@
+public abstract class Foldables {
+    
+    @SuppressWarnings("unchecked")
+    private static <T> T valueOf(Expression e, DataType to) {
```
I would remove this from ql since it's a trappy class. It uses `DataTypeConverter` which is QL specific and SQL would end up using that instead of `SqlDataTypeConverter`.
That is, `Foldables` is really reusable because of that and as such, I would just inline the methods.

matriv replied:
+1

costin reviewed (approved):
LGTM - however I would remove the use of `Foldables` classes either in this PR or a follow-up.

matriv reviewed (approved):
LGTM

astefan committed:
Be more explicit when folding and converting a foldable value, by
removing most of the code inside Foldables class.

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 52637_impl

astefan committed:
Clean up

astefan commented:
@matriv @costin another round, please? I've refactored `Foldables` by eliminating almost all methods in it, making the code clearer about values that are folded and **also converted**. `Foldables` wasn't only about folding foldable values.

matriv reviewed (approved):
LGTM

costin reviewed (approved):
LGTM
